### PR TITLE
Device Availability: Allow battery devices to report offline/online based on device.lastSeen

### PR DESCRIPTION
**Pull request by sjorge**
## Why
I have a few aqara sensors and sometimes one will lose connection and I don't always notice the stale data for a few hours/days.

## How?
Allow a `availability_lastseen_timeout` to be configure for battery powered devices.
```yaml
devices:
  '0x00158d00033ddfaa':
    friendly_name: serverroom/sensor
    retain: true
    debounce: 1
    availability_lastseen_timeout: 5400
```

Every time handleInterval is called, we compare the `device.lastSeen` with the current time. If the lastSeen was longer ago that `availability_lastseen_timeout` in **seconds** we mark the device as offline.

### Possible ways to improve this but I couldn't figure out how
- drop availability_lastseen_timeout and somehow set this timeout value in `zigbee-herdsman-convertors->devices.js` we then can check against that instead
- ~~allow this to be set for mains powered devices also, that way we do not cause extra traffic with the pings. (I can update the PR to include this, but not 100% this is a good idea?)~~
- **tests, I am not sure how to mock these**
    - device seen within timeout
    - device not seen within timeout
    - device not seen within timeout but returns by sending data

## Example
I set the timeout to 30 seconds, way to low as the aqara probes check in every 1h, more realistically would be 1.5h for example.

I let it go to offline, then I went and hit the little button to wake the sensor and force and report. The sensor then came back online.

```
zigbee2mqtt:info  2019-12-04 17:34:01: MQTT publish: topic 'zigbee2mqtt/serverroom/sensor/availability', payload 'offline'
zigbee2mqtt:error 2019-12-04 17:34:01: Device 'serverroom/sensor' was not seen for more than '30' seconds!
zigbee2mqtt:debug 2019-12-04 17:34:36: Received Zigbee message from 'serverroom/sensor', type 'attributeReport', cluster 'genBasic', data '{"65281":{"1":3005,"4":5032,"5":13,"6":[0,0],"10":0,"100":1774,"101":4263,"102":101863}}' from endpoint 1 with groupID 0
zigbee2mqtt:info  2019-12-04 17:34:36: MQTT publish: topic 'zigbee2mqtt/serverroom/sensor/availability', payload 'online'
zigbee2mqtt:debug 2019-12-04 17:34:36: Received Zigbee message from 'serverroom/sensor', type 'attributeReport', cluster 'genBasic', data '{"modelId":"lumi.weather"}' from endpoint 1 with groupID 0
zigbee2mqtt:info  2019-12-04 17:34:37: MQTT publish: topic 'zigbee2mqtt/serverroom/sensor', payload '{"battery":100,"voltage":3005,"temperature":17.74,"humidity":42.63,"pressure":1018.6,"linkquality":115}'
```

### CODE REVIEW

**Comment by runningman84** [comment]
Maybe there could be default timeouts based on the device type?

**Comment by sjorge** [comment]
The problem is not all devices have a fixed reporting,
e.g. Aqare temp sensor is ~1h (unless rapid change in one of the values), my TRV is... god knows what it varies from 10min to multiple hours...

See item too under improvements, for cyclic devices we should be able to define it somewhere so we can use that value unless overwritten or something. 

It’s always a balance between getting a functional thing merged or never getting anything merged because it will never be perfect or doubles.

**Comment by sjorge** [comment]
@Koenkk for the missing test coverage, I give it another go but I can't figure out how to mock a device that has the settings property set and then how to update the lastSeen attribute to between different steps :(

**Comment by sjorge** [comment]
I've removed the restriction of availability_lastseen_timeout only being used for battery devices. This can cut back on the ping traffic with for example tradfri bulbs that do status updates every few minutes.

**Comment by Koenkk** [comment]
For end devices I think we should use a static timeout of 24 hours, this should be fine for all end devices.

**Comment by sjorge** [comment]
I've been making a bit of progress with the tests too, not much though...

So if availability_lastseen_timeout is not specified for a battery device, assume `24 * 60 * 60`... I'll refactor this after I get some tests working against the current code.

**Comment by Kryzek** [comment]
> 
> 
> For end devices I think we should use a static timeout of 24 hours, this should be fine for all end devices.

I have some door and leakage sensors that do not say anything for ages:
image

Or it's just my device that is missing some configuration to report periodically (Wireless contact sensor (ZCTS-808) and Water leakage detector (ZWLD-100)) only updates I see is when these sensors see something happening (door opening and water and sensor).

**Comment by sjorge** [comment]
Already hit a few snags while testing the new commit :(

**Comment by sjorge** [comment]
```
zigbee2mqtt:debug 2019-12-07 18:12:58: Device 'serverroom/sensor' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:14:01: Device 'bedroom/radiator' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:14:16: Device 'bedroom/sensor' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:14:18: Device 'bedroom/night_light/remote' was last seen '1801' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:14:18: Device 'bedroom/desk_lamp/remote' was last seen '1801' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:02: Device 'bedroom/radiator' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:16: Device 'bedroom/sensor' was last seen '239' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:18: Device 'bedroom/night_light/remote' was last seen '1921' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:18: Device 'bedroom/desk_lamp/remote' was last seen '1921' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:19: Device 'serverroom/sensor' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:16:35: Device 'bedroom/motion' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:02: Device 'bedroom/radiator' was last seen '239' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:16: Device 'bedroom/sensor' was last seen '359' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:18: Device 'bedroom/night_light/remote' was last seen '2041' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:18: Device 'bedroom/desk_lamp/remote' was last seen '2041' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:19: Device 'serverroom/sensor' was last seen '239' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:18:35: Device 'bedroom/motion' was last seen '239' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:02: Device 'bedroom/radiator' was last seen '359' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:16: Device 'bedroom/sensor' was last seen '479' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:18: Device 'bedroom/night_light/remote' was last seen '2161' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:18: Device 'bedroom/desk_lamp/remote' was last seen '2161' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:19: Device 'serverroom/sensor' was last seen '359' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:20:35: Device 'bedroom/motion' was last seen '359' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:22:16: Device 'bedroom/sensor' was last seen '599' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:22:18: Device 'bedroom/night_light/remote' was last seen '2281' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:22:18: Device 'bedroom/desk_lamp/remote' was last seen '2281' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:22:19: Device 'serverroom/sensor' was last seen '479' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:22:35: Device 'bedroom/motion' was last seen '479' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:05: Device 'bedroom/radiator' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:16: Device 'bedroom/sensor' was last seen '719' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:18: Device 'bedroom/night_light/remote' was last seen '2401' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:18: Device 'bedroom/desk_lamp/remote' was last seen '2401' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:19: Device 'serverroom/sensor' was last seen '599' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:24:35: Device 'bedroom/motion' was last seen '599' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:05: Device 'bedroom/radiator' was last seen '119' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:16: Device 'bedroom/sensor' was last seen '839' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:18: Device 'bedroom/night_light/remote' was last seen '2521' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:18: Device 'bedroom/desk_lamp/remote' was last seen '2521' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:19: Device 'serverroom/sensor' was last seen '719' seconds ago.
zigbee2mqtt:debug 2019-12-07 18:26:51: Device 'bedroom/motion' was last seen '119' seconds ago.
```

@Koenkk aside from the tests it looks good, however like @Kryzek  I am not sure 24h by default is good, e.g. my remotes don't publish anything unless I use then I think. 

I will move one way so I don't use it to see if it is unavailable tomorrow or not.

**Comment by sjorge** [comment]
@Koenkk is there a way to filter device based on weather they have reporting setup or not?

I think skipping devices that do not have reporting should work in combination with a 24h default timeout.

**Comment by Koenkk** [comment]
This should be done now, all end devices which are not seen for more than 24 hours will be marked as unavailable now. @sjorge can you do a final check, then this can be merged.

**Comment by sjorge** [comment]
I'll update my local setup with the new commits and let it run for a few hours, also good news, the ikea on/off remote does send a daily update on the battery status... perhaps we should build in a tiny bit more margin of error like maybe 24h30m?

I also left some inline comments on the code.

File: lib/extension/deviceAvailability.js [hunk 1]
```diff
@@ -85,17 +97,31 @@ class DeviceAvailability extends BaseExtension {
...
             logger[level](`Failed to ping '${entity.name}'`);
         } finally {
-            this.setTimer(device);
+            this.setTimerPingable(device);
         }
     }
 
-    setTimer(device) {
+    async handleIntervalNotPingable(device) {
+        const ago = Date.now() - device.lastSeen;
```
Comment (sjorge) [inline lib/extension/deviceAvailability.js]: @Koenkk device.lastSeen is null on startup, should this not be moved below the check on line 107 ?

Comment (sjorge) [reply lib/extension/deviceAvailability.js]: Doens't seem to cause any errors though

**Comment by sjorge** [comment]
One thing of note, if zigbee2mqtt is restarted after a sensor goes misses a check-in but before the 24h time window has passed... it will never be reported as offline!

As we publish online at startup... device.lastSeen is null so it will never be checked.

**Comment by sjorge** [comment]
Aside from the few things mentioned above this seems to be working fine.

**Comment by Koenkk** [comment]
- I've changed the inactivity time to 25 hours (instead of 24 hours)
- `device.lastSeen` is indeed not persisted across restarts, but this is something that has to be done in zigbee-herdsman.

Thanks!

**Comment by sjorge** [comment]
👍 I'll open a issue against herdsman for that then.

**Pull request merged**